test/e2e/upgrade_test.go

```diff
@@ -0,0 +1,90 @@
+package e2e
+
+import (
+	"context"
+	"errors"
+	"math/rand"
+	"os"
+	"testing"
+	"time"
+
+	"github.com/celestiaorg/celestia-app/app"
+	"github.com/celestiaorg/celestia-app/app/encoding"
+	"github.com/celestiaorg/celestia-app/test/txsim"
+	"github.com/celestiaorg/knuu/pkg/knuu"
+	"github.com/stretchr/testify/require"
+)
+
+// This will only run tests within the v1 major release cycle
+const MajorVersion = 1
+
+func TestMinorVersionCompatibility(t *testing.T) {
+	if os.Getenv("E2E") != "true" {
+		t.Skip("skipping e2e test")
+	}
+
+	if os.Getenv("E2E_VERSIONS") == "" {
+		t.Skip("skipping e2e test: E2E_VERSION not set")
+	}
+
+	versionStr := os.Getenv("E2E_VERSIONS")
+	versions := ParseVersions(versionStr).FilterMajor(MajorVersion).FilterOutReleaseCandidates()
+	numNodes := 4
+	r := rand.New(rand.NewSource(seed))
+	t.Log("Running minor version compatibility test", "versions", versions)
+
+	testnet, err := New(t.Name(), seed)
+	require.NoError(t, err)
+	t.Cleanup(testnet.Cleanup)
+
+	// preload all docker images
+	preloader, err := knuu.NewPreloader()
+	require.NoError(t, err)
+	t.Cleanup(func() { _ = preloader.EmptyImages() })
+	for _, v := range versions {
+		err := preloader.AddImage(DockerImageName(v.String()))
+		require.NoError(t, err)
+	}
+
+	for i := 0; i < numNodes; i++ {
+		// each node begins with a random version within the same major version set
+		v := versions.Random(r).String()
+		t.Log("Starting node", "node", i, "version", v)
+		require.NoError(t, testnet.CreateGenesisNode(v, 10000000))
+	}
+
+	kr, err := testnet.CreateAccount("alice", 1e12)
+	require.NoError(t, err)
+
+	require.NoError(t, testnet.Setup())
+	require.NoError(t, testnet.Start())
+
+	sequences := txsim.NewBlobSequence(txsim.NewRange(200, 4000), txsim.NewRange(1, 3)).Clone(5)
+	sequences = append(sequences, txsim.NewSendSequence(4, 1000, 100).Clone(5)...)
+
+	errCh := make(chan error)
+	encCfg := encoding.MakeConfig(app.ModuleEncodingRegisters...)
+	opts := txsim.DefaultOptions().WithSeed(seed)
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	go func() {
+		errCh <- txsim.Run(ctx, testnet.GRPCEndpoints()[0], kr, encCfg, opts, sequences...)
+	}()
+
+	for i := 0; i < len(versions)*2; i++ {
+		// FIXME: skip the first node because we need them available to
+		// submit txs
+		if i%numNodes == 0 {
+			continue
+		}
```

Comment:
is skipping the first node also the mechanism that is testing that the versions are compatible?

Reply:
I don't quite understand. We need the first node to always be up because we're constantly submitting transactions. If txsim can't submit a transaction it exits.

Reply:
I see now that the test has been updated for each upgraded node to wait a few heights. I was curious what the mechanism was to ensure that different versions were running concurrently